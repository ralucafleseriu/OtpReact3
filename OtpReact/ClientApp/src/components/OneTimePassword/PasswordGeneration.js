import React, { useState, useEffect } from 'react';
import './OneTimePassword.css';
import PropTypes from 'prop-types';

const PasswordGeneration = ({ password, expirationDateTime, onGeneratePassword }) => {

  const [counter, setCounter] = useState();
  const [myInterval, setMyInterval] = useState();

  useEffect(() => {
    clearInterval(myInterval);
    setMyInterval()

    if (!expirationDateTime) return;

    setCounter(30);
    const id = setInterval(() => {
      if (expirationDateTime) {
        var c = Math.floor((expirationDateTime - new Date()) / 1000);
        setCounter(c)
      }
    }, 1000);
    setMyInterval(id);

    return () => {
      clearInterval(myInterval);
      setMyInterval()
    }
    // eslint-disable-next-line 
  }, [expirationDateTime]);

  function getPasswordStatus() {
    if (!counter) return null;
    if (counter <= 0) {
      return "Password expired.";
    }
    return `Valid for ${counter} seconds.`;
  }

  return (
    <div>
      <input value={password} readOnly />
      <button
        disabled={!!counter && counter > 0}
        onClick={() => onGeneratePassword()}
      >Generate one time password</button>
      <p>{getPasswordStatus()}</p>

    </div>
  );
}
PasswordGeneration.propTypes = {
  password: PropTypes.string,
  expirationDateTime: PropTypes.instanceOf(Date),
  onGeneratePassword: PropTypes.func
}
export default PasswordGeneration;
