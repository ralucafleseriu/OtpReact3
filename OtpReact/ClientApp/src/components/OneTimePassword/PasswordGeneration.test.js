import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import PasswordGeneration from './PasswordGeneration';

function renderPasswordGeneration(args) {
  const defaultProps = {
    password: "",
    expirationDateTime: Date(),
    onGeneratePassword: () => { }
  }
  const props = { ...defaultProps, ...args }
  return shallow(<PasswordGeneration {...props} />);
}

it('PasswordGeneration renders correctly', () => {
  const tree = renderer
    .create(<PasswordGeneration />);
  expect(tree).toMatchSnapshot();
});

it('renders fields correctly', () => {
  const handleGenerate = jest.fn();
  const wrapper = renderPasswordGeneration({
    onGeneratePassword: handleGenerate
  });
  expect(wrapper.find("input").length).toBe(1);

  // console.log(wrapper.find("button").debug());
  const generateButton = wrapper.find("button");
  expect(generateButton.length).toBe(1);
  expect(generateButton.text()).toBe("Generate one time password");

  generateButton.simulate("click");
  expect(handleGenerate.mock.calls.length).toBe(1);
});