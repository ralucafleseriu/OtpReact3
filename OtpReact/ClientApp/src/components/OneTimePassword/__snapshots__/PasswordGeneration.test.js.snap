// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordGeneration renders correctly 1`] = `
<div>
  <input
    readOnly={true}
  />
  <button
    disabled={false}
    onClick={[Function]}
  >
    Generate one time password
  </button>
  <p />
</div>
`;
